File: python/remove-nth-node-from-end-of-list.py
"""
Definition of ListNode
class ListNode(object):
    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""


class Solution:
    """
    @param head: The first node of linked list.
    @param n: An integer
    @return: The head of linked list.
    """

    def removeNthFromEnd(self, head, n):
        # write your code here
        stack = []

        node = head
        while node.next:
            stack.append(node)
            node = node.next

        for _ in range(n - 1):
            node = stack.pop()

        if not stack:
            return head.next
        prev = stack.pop()
        prev.next = node.next
        return head


# let quick pointer go for n first then continue


# def removeNthFromEnd(self, head, n):
#     res = ListNode(0)
#      res.next = head
#       tmp = res
#        for i in range(0, n):
#             head = head.next
#         while head != None:
#             head = head.next
#             tmp = tmp.next
#         tmp.next = tmp.next.next
#         return res.next

File: java/RemoveNthNodeFromEndOfList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    // https://leetcode.com/problems/remove-nth-node-from-end-of-list/
    // Iterative
    public ListNode removeNthFromEnd(ListNode head, int n) {
        if(head == null || n < 1) return head;
        ListNode node = new ListNode(0);
        node.next = head;
        
        ListNode slow = node;
        ListNode fast = node;
        // Moving fast pointer n places first
        for(int i = 0; i <= n; ++i) {
            fast = fast.next; 
        }
        
        // Then move both the pointer one step till fast pointer
        // reaches null. At this point the slow pointer will be 
        // at the right spot, just one node before the node that
        // needs to get deleted.
        while(fast != null) {
            fast = fast.next;
            slow = slow.next;
        }
        
        slow.next = slow.next.next;
        return node.next;
    }
    
    /* Recursive
    private boolean endReached = false;
    private int count = 0;
    private ListNode removeNthUtil(ListNode head, int n) {
        if(head == null) {
            endReached = true;
            return head;
        }
        head.next = removeNthUtil(head.next, n);
        
        if(endReached) {
            count++;
            if(count == n) {
               return head.next; 
            }
        }
        
        return head;
    }
    public ListNode removeNthFromEnd(ListNode head, int n) {
        if(head == null || n < 1) return head;
        return removeNthUtil(head, n);
    }
    */
}